File: solution/2181. Merge Nodes in Between Zeros/2181.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        vector<int> A;
        while (head) {
            if (head->val == 0)
                A.push_back(head->val);
            else {
                if (A.back() == 0)
                    A.push_back(head->val);
                else
                    A.back() += head->val;
            }
            head = head->next;
        }
        ListNode dummy;
        auto pre = &dummy;
        for (int x : A) {
            if (x == 0) continue;
            pre->next = new ListNode(x);
            pre = pre->next;
        }
        return dummy.next;
    }
};

File: solution/2181. Merge Nodes in Between Zeros/2181.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        p1, p2 = head, head.next
        
        while p2:
            if p2.val != 0:
                p1.val += p2.val
            else:
                p1.next = p2 if p2.next else None
                p1 = p2
            p2 = p2.next
        
        return head
    
